server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/mapping/MappingService.java

```diff
@@ -0,0 +1,246 @@
+      if (MappingHelper.size(mappingItem.extendedRange) > 0) {
+        result.add(mappingItem);
+      }
+
+      extendedDocumentLine += (range.getEnd().getLine() - originalDocumentLine);
+      originalDocumentLine = range.getEnd().getLine() + 1;
+
+      List<MappingItem> map = buildLocalityMap(textTransformations.getExtensions().get(range));
+      for (MappingItem item : map) {
+        Position start = item.extendedRange.getStart();
+        Position end = item.extendedRange.getEnd();
+        int size = end.getLine() - start.getLine() + 1;
+
+        start.setLine(extendedDocumentLine);
+        start.setCharacter(range.getStart().getCharacter());
+        end.setLine(extendedDocumentLine + size - 1);
+        item.extendedRange.setStart(start);
+        item.extendedRange.setEnd(end);
+        extendedDocumentLine += size;
+      }
+      result.addAll(map);
+    }
+
+    String[] originalLines = textTransformations.getText().split("\r?\n", -1);
+    int size = originalLines.length;
+
+    MappingItem last = new MappingItem(new Range(new Position(extendedDocumentLine, 0), new Position(extendedDocumentLine + size - 1, LINE_LEN)),
+        new Location(textTransformations.getUri(), new Range(new Position(originalDocumentLine, 0), new Position(originalDocumentLine + size - 1, LINE_LEN))), null);
+    result.add(last);
+
+    if (textTransformations.getReplacements().size() > 0) {
+      result = applyReplacements(result, textTransformations.getReplacements(), textTransformations.getText());
+    }
+
+    return result;
+  }
+
+  private MappingItem buildMappingItemForRange(String uri, Range range, int originalDocumentLine, int extendedDocumentLine) {
+    int size = range.getStart().getLine() - originalDocumentLine;
+    int charPos = range.getStart().getCharacter() - 1;
+    int originalEndLine = range.getStart().getLine();
+    int extendedEndLine = extendedDocumentLine + size;
+    if (charPos < 0) {
+      charPos = LINE_LEN;
+      originalEndLine -= 1;
+      extendedEndLine -= 1;
+    }
+    Location originalLocation = new Location(uri, new Range(new Position(originalDocumentLine, 0), new Position(originalEndLine, charPos)));
+    Range extended = new Range(new Position(extendedDocumentLine, 0), new Position(extendedEndLine, charPos));
+
+    return new MappingItem(extended, originalLocation, null);
+  }
+
+  private static List<MappingItem> applyReplacements(List<MappingItem> localityMap, Map<Range, String> replacements, String text) {
+    localityMap = new ArrayList<>(localityMap);
+    List<Range> ranges = new LinkedList<>(replacements.keySet());
+    ranges = ranges.stream().filter(r -> affectsToMapping(r, replacements.get(r))).collect(Collectors.toList());
+    ranges.sort(Comparator.comparingInt(e -> e.getStart().getLine()));
+
+    for (Range range : ranges) {
+      if (MappingHelper.size(range) <= 1) {
+        continue;
+      }
+
+      String[] lines =  text.split("\\r?\\n");
```

Comment:
Just a doubt. Can System.lineSeparator help us here?

Reply:
It is regex, so maybe not :-)